### PR TITLE
tests: functional: switch to rsync for directory diffs

### PR DESCRIPTION
**Pull request by cyphar**
This has been split off from #12209 so that it can be reviewed separately and the ZFS test failures (which appear to be real bugs in ZFS or the test suite) can be resolved separately to the renameat2 patchset.

While "diff -r" is the most straightforward way of comparing directory
trees for differences, it has two major issues:

 * File metadata is not compared, which means that subtle bugs may be
   missed even if a test is written that exercises the buggy behaviour.
 * diff(1) doesn't know how to compare special files -- it assumes they
   are always different, which means that a test using diff(1) on
   special files will always fail (resulting in such tests not being
   added).

rsync can be used in a very similar manner to diff (with the -ni flags),
but has the additional benefit of being able to detect and resolve many
more differences between directory trees. In addition, rsync has a
standard set of features and flags while diffs feature set depends on
whether you're using GNU or BSD binutils.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
@cyphar the causes of the failures here is that `rsync` is detecting a change in mtime.  The replay tests are a good example, the ZIL record only contains the create time which is set correctly, but the current mtime is used during replay.  We could change this in the code, but since it's the expected long standing behavior I'd rather update the tests to handle it.

I'd suggest pulling commit 06947ea2c1f4384281cf17e4484e8e2ea887d96d in to this PR (or something like it) to resolve these failures.  Locally the failing tests pass for me with this change applied.

**Comment by cyphar** [comment]
> the causes of the failures here is that rsync is detecting a change in mtime. The replay tests are a good example, the ZIL record only contains the create time which is set correctly, but the current mtime is used during replay. We could change this in the code, but since it's the expected long standing behavior I'd rather update the tests to handle it.

Ah okay, I assumed that replay would recover the mtime as well as the ctime. That makes a lot of sense. I am a bit confused why rsync reports that there are different xattrs under FreeBSD but I'll apply the commit you mentioned  (or try to figure out a slightly cleaner way of doing it -- I think rsync should have a flag that ignores mtimes).

**Comment by freqlabs** [comment]
The failure is reproducible. listing the contents of each dir right before the rsync comparison in slog_replay_fs_001 is a bit more readable than the itemized changes for me. I notice the size differences, but I don't know why they're different.
original:
```
total 10273
drwxr-xr-x  3 root  wheel       19 Sep 28 19:51 .
drwxr-xr-x  3 root  wheel        3 Sep 28 19:51 ..
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 b
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 c
lrwxr-xr-x  1 root  wheel       18 Sep 28 19:51 d -> /testpool/testfs/c
-rw-r--r--  2 root  wheel        0 Sep 28 19:51 e
-rw-r--r--  2 root  wheel        0 Sep 28 19:51 f
-rw-r--r--  1 root  wheel   262144 Sep 28 19:51 holes.1
-rw-r--r--  1 root  wheel   524288 Sep 28 19:51 holes.2
-rw-r--r--  1 root  wheel  1048576 Sep 28 19:51 holes.3
-rw-r--r--  1 root  wheel  8388608 Sep 28 19:51 large
-rw-r--r--  1 root  wheel  1048576 Sep 28 19:51 link_and_unlink
-rw-r--r--  1 root  wheel     8192 Sep 28 19:51 payload
-r-xrw-rwx  1 root  wheel        0 Nov 27  2013 setattr
-rw-r--r--  1 root  wheel      512 Sep 28 19:51 small_file
-rw-r--r--  1 root  wheel        1 Sep 28 19:51 sync
-rw------T  1 root  wheel        0 Sep 28 19:51 truncated_file
drwxr-xr-x  2 root  wheel        2 Sep 28 19:51 xattr.dir
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 xattr.file
```
copy:
```
total 11172
drwxr-xr-x  3 root  wheel      512 Sep 28 19:51 .
drwxr-xr-x  3 root  wheel      512 Sep 28 19:51 ..
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 b
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 c
lrwxr-xr-x  1 root  wheel       18 Sep 28 19:51 d -> /testpool/testfs/c
-rw-r--r--  2 root  wheel        0 Sep 28 19:51 e
-rw-r--r--  2 root  wheel        0 Sep 28 19:51 f
-rw-r--r--  1 root  wheel   262144 Sep 28 19:51 holes.1
-rw-r--r--  1 root  wheel   524288 Sep 28 19:51 holes.2
-rw-r--r--  1 root  wheel  1048576 Sep 28 19:51 holes.3
-rw-r--r--  1 root  wheel  8388608 Sep 28 19:51 large
-rw-r--r--  1 root  wheel  1048576 Sep 28 19:51 link_and_unlink
-rw-r--r--  1 root  wheel     8192 Sep 28 19:51 payload
-r-xrw-rwx  1 root  wheel        0 Nov 27  2013 setattr
-rw-r--r--  1 root  wheel      512 Sep 28 19:51 small_file
-rw-r--r--  1 root  wheel        1 Sep 28 19:51 sync
-rw------T  1 root  wheel        0 Sep 28 19:51 truncated_file
drwxr-xr-x  2 root  wheel      512 Sep 28 19:51 xattr.dir
-rw-r--r--  1 root  wheel        0 Sep 28 19:51 xattr.file
```

File: tests/zfs-tests/include/libtest.shlib [hunk 1]
```diff
@@ -4272,3 +4272,48 @@ function wait_for_children #children
...
+function directory_diff # dir_a dir_b [ignore_mtime]
+{
+	dir_a="$1"
+	dir_b="$2"
+	typeset ignore_mtime="${3:-0}"
+
+	extra_args=()
+
+	# NOTE: Quite a few rsync builds do not support --crtimes which would be
+	# necessary to verify that creation times are being maintained properly. So
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib:4293]:
```suggestion
	# necessary to verify that creation times are being maintained properly.
```

**Review by behlendorf — COMMENTED** [review]
> I notice the size differences, but I don't know why they're different.

Interestingly after spot checking a few of the builders I only saw the size differences on FreeBSD.  Most of this code is common so that's a bit surprising, I'm not sure what's going on there either.

> or try to figure out a slightly cleaner way of doing it

I'm all for it!  Although I couldn't find a way myself, it also looks like perhaps `--no-times` isn't quite sufficient based on the latest CI results.

**Comment by behlendorf** [comment]
@cyphar to move this forward we're going to need to somehow suppress the mtime/crtime warnings.  Those are expected.  The size warnings on FreeBSD for the directories are interesting but not harmful so let's update the test for now to ignore them as well.

**Comment by freqlabs** [comment]
Agreed, the test change seems good and the failure it exposes can be investigated separately.

**Comment by cyphar** [comment]
Sorry, I was busy and only just got back to this now. I've rebased it and cleaned up the ignore_mtimes usage so it's more clear what's going on in the handful of places that use it. I haven't yet fixed the directory size issue (I want to take a look at the error logs before trying to write some code to ignore it).

File: tests/zfs-tests/include/libtest.shlib [hunk 2]
```diff
@@ -4303,3 +4303,75 @@ function wait_for_children #children
...
+	# Unfortunately because of this we cannot use it unconditionally but we can
+	# check if this rsync build supports it and use it then. This check is
+	# based on the same check in the rsync test suite (testsuite/crtimes.test).
+	#
+	# We check ctimes even with zil_replay=1 because the ZIL does store
+	# creation times.
+	if ( rsync --version | grep "[, ] crtimes" >/dev/null ); then
```
Comment (nabijaczleweli) [inline tests/zfs-tests/include/libtest.shlib:4336]: `grep -q`?

**Comment by cyphar** [comment]
@behlendorf We are now ignoring all timestamp errors (Linux tests still seem to fail but they seem unrelated since those tests don't use `directory_diff` and there are no `directory_diff` failures under Linux). However (as I mentioned above), it appears that on FreeBSD, xattrs are being changed after a ZIL replay:

```
13:21:25.81 NOTE: Verify working set diff:
13:21:25.81 note: this rsync package does not support --crtimes
13:21:25.87 .f........x. xattr.file
13:21:25.87 .d........x. xattr.dir/
13:21:25.88 
13:21:25.88 ERROR: replay_directory_diff /mnt/testdir/copy /testpool/testfs exited 1
```

I will need to investigate this a bit more, but this seems to be something we shouldn't ignore? If you'd like me to ignore xattr errors specifically for this test, I can do that too. I'll spin up a FreeBSD VM to figure out what the difference is.

**Comment by cyphar** [comment]
@behlendorf @freqlabs We are now ignoring all timestamp errors (Linux tests still seem to fail but they seem unrelated since those tests don't use `directory_diff` and there are no `directory_diff` failures under Linux). However (as I mentioned above), it appears that on FreeBSD, xattrs are being changed after a ZIL replay:

```
13:21:25.81 NOTE: Verify working set diff:
13:21:25.81 note: this rsync package does not support --crtimes
13:21:25.87 .f........x. xattr.file
13:21:25.87 .d........x. xattr.dir/
13:21:25.88 
13:21:25.88 ERROR: replay_directory_diff /mnt/testdir/copy /testpool/testfs exited 1
```

I will need to investigate this a bit more, but this seems to be something we shouldn't ignore? If you'd like me to ignore xattr errors specifically for this test, I can do that too. I'll spin up a FreeBSD VM to figure out what the difference is.

File: tests/zfs-tests/include/libtest.shlib (hunk 2 above)
Comment (cyphar) [reply tests/zfs-tests/include/libtest.shlib:4336]: Good point, fixed.

**Comment by cyphar** [comment]
I just tried to reproduce the xattr issue in a FreeBSD 13-RELEASE VM, and I couldn't get it to work.

```
% truncate 10G disk
% mdconfig -a -t vnode -f disk
md0
% zpool create tank md0
% zfs create tank/fs
% sync if=/dev/zero of=/tank/fs/sync bs=1 count=1 conv=fsync,fdatasync
...
% zpool freeze tank
% touch /tank/fs/xattr.file
% mkdir /tank/fs/xattr.dir
% setextattr user fileattr HelloWorldFile /tank/fs/xattr.file
% setextattr user fileattr HelloWorldDir /tank/fs/xattr.dir
% setextattr user tmpattr HelloWorld /tank/fs/xattr.file
% setextattr user tmpattr HelloWorld /tank/fs/xattr.dir
% rmextattr user tmpattr /tank/fs/xattr.file
% rmextattr user tmpattr /tank/fs/xattr.dir
% rsync -aAHX /tank/fs/ /tankfs-backup/
% zfs unmount tank/fs
% zpool export tank
% zpool import -f tank
% rsync -ni -aAHX /tank/fs/ /tankfs-backup/
.d..t....... ./
```
I'm really not sure why this is failing within the tests -- the xattrs are actually identical.

This was done with the built-in ZFS on FreeBSD -- I'm not familiar enough with FreeBSD to compile a custom kernel module to try the latest version of OpenZFS. If someone else could take a look at this, I'd really appreciate it.

**Comment by freqlabs** [comment]
Sure, I'll take a look.

**Comment by behlendorf** [comment]
> Linux tests still seem to fail but they seem unrelated since those tests don't use directory_diff and there are no directory_diff failures under Linux

Perhaps this comment was for a previous run, but the [latest CI results](http://build.zfsonlinux.org/builders/CentOS%20Stream%208%20x86_64%20%28TEST%29/builds/4155/steps/shell_4/logs/summary) show the Linux bots failing on these tests which were modified to use `directory-diff`.  For reference we do still get CI failures, but normally it's just one or two of the bots.

```
    FAIL cli_root/zfs_snapshot/zfs_snapshot_009_pos (expected PASS)
    FAIL rsend/recv_dedup (expected PASS)
    FAIL slog/slog_replay_fs_001 (expected PASS)
    FAIL slog/slog_replay_fs_002 (expected PASS)
```

**Comment by cyphar** [comment]
@behlendorf Ah, I was looking at the Debian tests (which aren't failing in any of the `directory_diff` users). But they are failing on CentOS -- I suspect it's got to do with SELinux because all of the differences are xattr-related. This might be a bit tricky to fix, let me see if I can restrict xattr comparisons in rsync to just `user.*`...

File: tests/zfs-tests/include/libtest.shlib [hunk 3]
```diff
@@ -4303,3 +4303,80 @@ function wait_for_children #children
...
+	# do not match there is a "c" entry in one of the columns).
+	if ( rsync --version | grep -q "[, ] crtimes" >/dev/null ); then
+		args+=("--crtimes")
+	else
+		echo "NOTE: This rsync package does not support --crtimes (-N)."
+	fi
+
+	# If we are testing a ZIL replay, we need to ignore timestamp changes.
+	# Unfortunately --no-times doesn't do what we want -- it will still tell if
+	# you the timestamps don't match but rsync will set the timestamps to the
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib:4349]: typo: flip "if" and "you" here, "it will still tell you if the timestamps...".

It's nice to see all the corner cases clearly explained in this function.

File: tests/zfs-tests/include/libtest.shlib [hunk 4]
```diff
@@ -4303,3 +4303,80 @@ function wait_for_children #children
...
+	# doesn't compare file metadata, and cannot handle special files).
+	#
+	# Also make sure to filter out non-user.* xattrs when comparing. On
+	# SELinux-enabled systems the copied tree will probably have different
+	# SELinux labels.
+	args=("-nicaAHX" '--filter=-x! user.*' "--delete")
+
+	# NOTE: Quite a few rsync builds do not support --crtimes which would be
+	# necessary to verify that creation times are being maintained properly. So
+	# Unfortunately because of this we cannot use it unconditionally but we can
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib:4334]: Looks like you still want to drop "So" from the previous line, or uncapitalize "Unfortunately".

**Comment by cyphar** [comment]
@behlendorf After many many many test fixes, all the tests are passing now. :smile_cat:

**Review by behlendorf — APPROVED** [review]
Looks good.  Thanks for sorting out all of those additional test case fixes so we can get this integrated.  You probably already saw, but the two FreeBSD failures are known unrelated issues so no need to worry about this.

```
    FAIL cli_root/zfs_receive/receive-o-x_props_override (expected PASS)
    FAIL largest_pool/largest_pool_001_pos (expected PASS)
```
